sunpy/util/decorators.py

```diff
@@ -325,3 +325,53 @@ def inner_f(*args, **kwargs):
             return f(**kwargs)
         return inner_f
     return deprecate_positional_args
+
+
+_NOT_FOUND = object()
+
+
+def cached_property_based_on(meth):
+    """
+    A decorator to cache the value of a property based on the output of a
+    different callable.
```

Comment:
Wouldn't making the input here a attribute / property and not a callable be a little clearer? Is the reason it's a method to deal with scoping issues when defining the class?

An alternative could be to do it by name rather than passing in the actual method
```
@cached_property_based_on("_meta_hash")
```
which then does a `getattr` on the object?

Reply:
The reason I did this is because handing over an object (unless it's handing it by reference and not value?) doesn't work, as there needs to be some way to detect changes. Didn't think of doing `getattr` though, and agree that is a bit of a nicer solution, so will update it.

Reply:
(I would expect a property to work as it's really a callable underneath, so it's effectively by reference. A generic attribute may or may not work depending I guess :+1: )

Reply:
I think I've updated this to what you meant now?